File: 0649-dota2-senate/0649-dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        senators = defaultdict(list)
        for i, ch in enumerate(senate):
            senators[ch].append(i)

        i = 0
        while True:
            if not senators["R"]:
                return "Dire"
            elif not senators["D"]:
                return "Radiant"
            if senators["R"][0] == i:
                senators["R"].append(senators["R"].pop(0))
                senators["D"].pop(0)
            elif senators["D"][0] == i:
                senators["D"].append(senators["D"].pop(0))
                senators["R"].pop(0)
            i = (i + 1) % len(senate)

File: 0649-dota2-senate/0649-dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        unordered_map<char, queue<int>> senators;
        for (int i = 0; i < senate.size(); ++i) {
            senators[senate[i]].push(i);
        }

        int i = 0;
        while (true) {
            if (senators['R'].empty()) {
                return "Dire";
            } else if (senators['D'].empty()) {
                return "Radiant";
            }
            if (senators['R'].front() == i) {
                senators['R'].pop();
                senators['D'].pop();
                senators['R'].push(i);
            } else if (senators['D'].front() == i) {
                senators['D'].pop();
                senators['R'].pop();
                senators['D'].push(i);
            }
            i = (i + 1) % senate.size();
        }
    }
};